packages/amplify-prompts/src/progressbars/spinner.ts

```diff
@@ -3,64 +3,83 @@
+  private frames: string[];
+  private timer!: ReturnType<typeof setTimeout>;
+  private prefixText: string;
+  private terminal: Terminal | null;
+  private refreshRate: number;
 
-    constructor(text : string) {
-      this.frameCount = 0;
-      this.frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
-      this.prefixText = text;
-      this.terminal = new Terminal();
-      this.refreshRate = 50;
-    }
+  constructor() {
+    this.frameCount = 0;
+    this.frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
+    this.prefixText = '';
+    this.refreshRate = 50;
+    this.terminal = null;
+  }
 
-    /**
-     * Render function
-     */
-    render() : void {
-      if (this.timer) {
-        clearTimeout(this.timer);
-      }
-      const lines = [{
-        renderString: `${this.frames[this.frameCount]} ${this.prefixText}`,
-        color: '',
-      }];
-      this.frameCount = ++this.frameCount % this.frames.length;
+  /**
+   * Render function
+   */
+  private render(): void {
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    const lines = [{
+      renderString: `${this.frames[this.frameCount]} ${this.prefixText}`,
+      color: '',
+    }];
+    this.frameCount = ++this.frameCount % this.frames.length;
+    if (this.terminal) {
       this.terminal.writeLines(lines);
-      this.timer = setTimeout(() => this.render(), this.refreshRate);
     }
+    this.timer = setTimeout(() => this.render(), this.refreshRate);
+  }
 
-    /**
-     * Starts a spinner and calls render function.
-     */
-    start() : void {
-      this.prefixText = this.prefixText.replace('\n', '');
-      this.terminal.cursor(false);
-      this.render();
+  /**
+   * Starts a spinner and calls render function.
+   */
+  start(text: string | null): void {
+    if (!this.terminal) {
+      this.terminal = new AmplifyTerminal();
     }
+    this.prefixText = text ? text.replace('\n', '') : this.prefixText;
+    this.terminal.cursor(false);
+    this.render();
+  }
 
-    /**
-     * Stops the spinner
-     */
-    stop(text : string | null, success = true) : void {
-      if (text) {
-        const lines : StringObj[] = [{
-          renderString: text,
-          color: success ? 'green' : 'red',
-        }];
+  /**
+   * Reset Spinner message
+   */
+  resetMessage(text: string | null): void {
+    if (!this.terminal) {
+      this.start(text);
+      return;
+    }
+    this.prefixText = text ? text.replace('\n', '') : this.prefixText;
+  }
 
-        clearTimeout(this.timer);
-        this.terminal.writeLines(lines);
-      }
+  /**
+   * Stops the spinner
+   */
+  stop(text? : string | null, success = true): void {
+    if (!this.terminal) {
+      return;
     }
+    const lines : StringObj[] = [{
+      renderString: text || '',
+      color: success ? 'green' : 'red',
+    }];
+
+    clearTimeout(this.timer);
+    this.terminal.writeLines(lines);
+    this.terminal.cursor(true);
+    this.terminal = null;
```

Comment:
should a spinner be single usage? Does TS have a concept of disposable (aka closeable) ?
how do we prevent bad usage? I.e. what happens if multiple async stacks share reference to this?

Reply:
I think that might make sense, I've added this as a backlog item to refactor.

I'm not aware of such a concept in TS.

Preventing bad usage is up to the caller. This is similar to how we use `ora`.